pkg/models/iam/im/authenticator.go

```diff
@@ -0,0 +1,228 @@
+)
+
+type PasswordAuthenticator interface {
+	Authenticate(username, password string) (authuser.Info, error)
+}
+
+type passwordAuthenticator struct {
+	ksClient          kubesphere.Interface
+	userLister        iamv1alpha2listers.UserLister
+	loginRecordLister iamv1alpha2listers.LoginRecordLister
+	options           *authoptions.AuthenticationOptions
+}
+
+func NewPasswordAuthenticator(ksClient kubesphere.Interface,
+	userLister iamv1alpha2listers.UserLister,
+	loginRecordLister iamv1alpha2listers.LoginRecordLister,
+	options *authoptions.AuthenticationOptions) PasswordAuthenticator {
+	return &passwordAuthenticator{
+		ksClient:          ksClient,
+		userLister:        userLister,
+		loginRecordLister: loginRecordLister,
+		options:           options}
+}
+
+func (im *passwordAuthenticator) Authenticate(username, password string) (authuser.Info, error) {
+
+	user, err := im.searchUser(username)
+
+	if err != nil {
+		// internal error
+		if !errors.IsNotFound(err) {
+			klog.Error(err)
+			return nil, err
+		}
+	}
+
+	providerOptions, ldapProvider := im.getLdapProvider()
+
+	// no identity provider
+	if user == nil && (providerOptions == nil || providerOptions.MappingMethod != oauth.MappingMethodAuto) {
+		return nil, AuthFailedIncorrectPassword
+	}
+
+	// able to login using the locally principal admin account and password in case of a disruption of LDAP services.
+	if ldapProvider != nil && username != constants.AdminUserName {
+		if providerOptions.MappingMethod == oauth.MappingMethodLookup &&
+			(user == nil || user.Labels[iamv1alpha2.IdentifyProviderLabel] != providerOptions.Name) {
+			klog.Error(AuthFailedIdentityMappingNotMatch)
+			return nil, AuthFailedIdentityMappingNotMatch
+		}
+		if providerOptions.MappingMethod == oauth.MappingMethodAuto &&
+			user != nil && user.Labels[iamv1alpha2.IdentifyProviderLabel] != providerOptions.Name {
+			klog.Error(AuthFailedIdentityMappingNotMatch)
+			return nil, AuthFailedIdentityMappingNotMatch
+		}
+
+		authenticated, err := ldapProvider.Authenticate(username, password)
+		if err != nil {
+			if ldap.IsErrorWithCode(err, ldap.LDAPResultInvalidCredentials) ||
+				ldap.IsErrorWithCode(err, ldap.LDAPResultNoSuchObject) {
+				klog.Error(err)
+				return nil, AuthFailedIncorrectPassword
+			} else {
+				klog.Error(err)
+				return nil, err
+			}
+		}
+
+		if authenticated != nil && user == nil {
+			authenticated.Labels = map[string]string{iamv1alpha2.IdentifyProviderLabel: providerOptions.Name}
+			if authenticated, err = im.ksClient.IamV1alpha2().Users().Create(authenticated); err != nil {
+				klog.Error(err)
+				return nil, err
+			}
+		}
+
+		if authenticated != nil {
+			return &authuser.DefaultInfo{
+				Name: authenticated.Name,
+				UID:  string(authenticated.UID),
+			}, nil
+		}
+	}
+
+	if limitExceed, err := im.authRateLimitExceeded(user); err != nil {
+		klog.Error(err)
+		return nil, err
+	} else if limitExceed {
+		klog.V(4).Info(AuthRateLimitExceeded)
```

Comment:
Error log should be more informative, please log user as well

Reply:
fixed